src/Tactic/Rewrite.agda

```diff
@@ -0,0 +1,222 @@
+-- 'rewrite' during equational reasoning. This allows us to write very
+-- succinct proofs:
+--
+-- example : ∀ m n → m ≡ n → suc (suc (m + 0)) + m ≡ suc (suc n) + (n + 0)
+-- example m n eq = begin
+--     suc (suc (m + 0)) + m
+--   ≡⟨ rw (+-identityʳ m) ⟩
+--     suc (suc m) + m
+--   ≡⟨ rw eq ⟩
+--     suc (suc n) + n
+--   ≡˘⟨ rw (+-identityʳ n) ⟩
+--     suc (suc n) + (n + 0)
+--   ∎
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Tactic.Rewrite where
+
+open import Function using (_$_)
+
+open import Data.Bool.Base            using (true; false; if_then_else_; _∧_)
+open import Data.Char.Base   as Char  using (toℕ)
+open import Data.Float.Base  as Float using (_≡ᵇ_)
+open import Data.List.Base   as List  using ([]; _∷_)
+open import Data.Maybe.Base  as Maybe using (Maybe; just; nothing)
+open import Data.Nat.Base    as Nat   using (ℕ; zero; suc; _≡ᵇ_; _+_)
+open import Data.Unit.Base            using (⊤)
+open import Data.Word.Base   as Word  using (toℕ)
+open import Data.Product
+
+open import Relation.Binary.PropositionalEquality as Eq using (_≡_; refl; cong)
+
+-- 'Data.String.Properties' defines this via 'Dec', so let's use the builtin
+-- for maximum speed.
+import Agda.Builtin.String as String renaming (primStringEquality to _≡ᵇ_)
+
+open import Reflection
+open import Reflection.Abstraction
+open import Reflection.AlphaEquality        as Alpha
+open import Reflection.Argument             as Arg
+open import Reflection.Argument.Information as ArgInfo
+open import Reflection.Argument.Visibility  as Visibility
+open import Reflection.Meta                 as Meta
+open import Reflection.Name                 as Name
+open import Reflection.Term                 as Term
+
+open import Reflection.TypeChecking.Monad.Syntax
+
+----------------------------------------------------------------------
+-- Utilities
+----------------------------------------------------------------------
+
+private
+  -- Descend past a variable.
+  var-descend : ℕ → ℕ → ℕ
+  var-descend ϕ x = if ϕ Nat.≤ᵇ x then suc x else x
+
+  -- Descend a variable underneath pattern variables.
+  pattern-descend : ℕ → Pattern → (Pattern × ℕ)
+  patterns-descend : ℕ → Args Pattern → (Args Pattern × ℕ)
+
+  pattern-descend ϕ (con c ps) =
+    let (ps' , ϕ') = patterns-descend ϕ ps
+    in (con c ps' , ϕ')
+  pattern-descend ϕ (dot t)    = (dot t) , ϕ
+  pattern-descend ϕ (var x)    = var (var-descend ϕ x) , suc ϕ
+  pattern-descend ϕ (lit l)    = (lit l) , ϕ
+  pattern-descend ϕ (proj f)   = (proj f) , ϕ
+  pattern-descend ϕ (absurd x) = absurd (var-descend ϕ x) , suc ϕ
+
+  patterns-descend ϕ ((arg i p) ∷ ps) =
+    let (p' , ϕ') = pattern-descend ϕ p
+        (ps' , ϕ'') = patterns-descend ϕ' ps
+    in (arg i p ∷ ps' , ϕ'')
+  patterns-descend ϕ []       =
+    [] , ϕ
+
+  -- Helper for constructing applications of 'cong'
+  `cong : Term → Term → Term
```

Comment:
So we should probably add this and `endpoints` to a new file `Relation.Binary.PropositionalEquality.Reflection`? That way they'd be reusable by other tactics in future?

I guess `endpoints` could take the error message string to throw as an extra argument?